import { createAsyncThunk } from "@reduxjs/toolkit";
import { FirebaseError } from "firebase/app";
import { createUserWithEmailAndPassword, signInWithEmailAndPassword, User } from "firebase/auth";
import { auth } from "../../config/firebase";

export const postUser = createAsyncThunk<User, { email: string; password: string }, { rejectValue: string }>(
  "user/createUser",
  async ({ email, password }, thunkAPI) => {
    try {
      const userCredential = await createUserWithEmailAndPassword(auth, email, password);

      return userCredential.user.toJSON() as User;
    } catch (error) {
      if (error instanceof FirebaseError) {
        return thunkAPI.rejectWithValue(error.message);
      }
      return thunkAPI.rejectWithValue("N책gontin gick fel vid skapandet av anv채ndare. Kontakta v책r support om felet kvartst책r.");
    }
  }
);

export const signInUser = createAsyncThunk<User, { email: string; password: string }, { rejectValue: string }>(
  "user/signInUser",
  async ({ email, password }, thunkAPI) => {
    try {
      const userCredentials = await signInWithEmailAndPassword(auth, email, password);

      return userCredentials.user.toJSON() as User;
    } catch (error) {
      if (error instanceof FirebaseError) {
        return thunkAPI.rejectWithValue(error.message);
      }
      return thunkAPI.rejectWithValue("Error logging in user");
    }
  }
);
